From: Barry Song <v-songbaohua@oppo.com>
Subject: mm-use-folio_add_new_anon_rmap-if-folio_test_anonfolio==false-fix-2
Date: Sat, 22 Jun 2024 15:14:53 +1200

For new anon(!anon), there's a possibility that multiple concurrent
threads might execute "if (!anon) folio_add_new_anon_rmap()" in parallel. 
In such cases, the threads should lock the folio before executing this
sequence.  We use VM_WARN_ON_FOLIO() to verify if this condition holds
true.

Link: https://lkml.kernel.org/r/20240622032002.53033-1-21cnbao@gmail.com
Signed-off-by: Barry Song <v-songbaohua@oppo.com>
Suggested-by: David Hildenbrand <david@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/memory.c   |    1 +
 mm/swapfile.c |    1 +
 2 files changed, 2 insertions(+)

--- a/mm/memory.c~mm-use-folio_add_new_anon_rmap-if-folio_test_anonfolio==false-fix-2
+++ a/mm/memory.c
@@ -4346,6 +4346,7 @@ check_folio:
 		 * here, we have to be careful.
 		 */
 		VM_WARN_ON_ONCE(folio_test_large(folio));
+		VM_WARN_ON_FOLIO(!folio_test_locked(folio), folio);
 		folio_add_new_anon_rmap(folio, vma, address, rmap_flags);
 	} else {
 		folio_add_anon_rmap_ptes(folio, page, nr_pages, vma, address,
--- a/mm/swapfile.c~mm-use-folio_add_new_anon_rmap-if-folio_test_anonfolio==false-fix-2
+++ a/mm/swapfile.c
@@ -1923,6 +1923,7 @@ static int unuse_pte(struct vm_area_stru
 		 */
 		if (!folio_test_anon(folio)) {
 			VM_WARN_ON_ONCE(folio_test_large(folio));
+			VM_WARN_ON_FOLIO(!folio_test_locked(folio), folio);
 			folio_add_new_anon_rmap(folio, vma, addr, rmap_flags);
 		} else {
 			folio_add_anon_rmap_pte(folio, page, vma, addr, rmap_flags);
_
